hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationUtil.java

```diff
@@ -54,6 +59,32 @@
     private SerializationUtil() {
     }
 
+    /**
+     * Checks that the {@code object} implements {@link Serializable} and is
+     * correctly serializable by actually trying to serialize it. This will
+     * reveal some non-serializable field early.
+     *
+     * @param object     object to check
+     * @param objectName object description for the exception
+     * @throws IllegalArgumentException if {@code object} is not serializable
+     */
+    public static void checkSerializable(Object object, String objectName) {
```

Comment:
I'm pretty worried about this because AFAIK on the happy path we can end up with an extra serialization of the slowest kind.

Reply:
You mean the part where we actually serialise - `os.writeObject(object);`? I guess we can remove it but it's not that bad. The method is called only once, when creating a comparator and it's only used for creating comparators and no other objects. For example:
```java
static <T, U extends Comparable<? super U>> ComparatorEx<T> comparing(Function<? super T, ? extends U> toKeyFn) {
        checkNotNull(toKeyFn, "toKeyFn");
        checkSerializable(toKeyFn, "toKeyFn");
        return (left, right) -> toKeyFn.apply(left).compareTo(toKeyFn.apply(right));
    }
```